File: examples/lmdif.jl
using LsqFit, Calculus

# a simple wrapper for Levenberg-Marquardt with finite differencing for the Jacobian
lmdif(f, x0; show_trace::Bool=false) =
 LsqFit.levenberg_marquardt(f, jacobian(f), x0, show_trace=show_trace)

# custom MEX-like wrapper
using MATLAB: jvalue
lmdif(args::Vector{MATLAB.MxArray}) =
 lmdif(jvalue(args[1]), jvalue(args[2]), show_trace=(length(args)>2 ? jvalue(args[3]) : false))

File: examples/lmdif.m
% Levenberg-Marquardt solver with finite differencing for the Jacobian
function x = lmdif(f, x0)

% first time through, load the julia function
persistent loaded;
if isempty(loaded)
  jl_file = fullfile(fileparts(mfilename('fullpath')), 'lmdif.jl');
  jl.include(jl_file);
  loaded = true;
end

sln = jl.callkw(2, 'lmdif', f, x0, 'show_trace', true);

if ~(sln.x_converged || sln.f_converged || sln.g_converged)
  throw('lmdif failed to converge');
end

x = sln.minimizer;

end
